src/transaction/log_replication.cpp

```diff
@@ -1,173 +1,264 @@
+	m_parallel_replication_redo->wait_for_termination_and_stop_execution ();
+      }
+
+    log_zip_free_data (m_undo_unzip);
+    log_zip_free_data (m_redo_unzip);
+  }
+
+  void
+  replicator::redo_upto_nxio_lsa (cubthread::entry &thread_entry)
+  {
+    thread_entry.tran_index = LOG_SYSTEM_TRAN_INDEX;
+
+    while (true)
+      {
+	const log_lsa nxio_lsa = log_Gl.append.get_nxio_lsa ();
+	if (m_redo_lsa < nxio_lsa)
+	  {
+	    redo_upto (thread_entry, nxio_lsa);
+	  }
+	else
+	  {
+	    assert (m_redo_lsa == nxio_lsa);
+	    break;
+	  }
+      }
+  }
+
+  void
+  replicator::wait_parallel_replication_idle ()
+  {
+    if (m_parallel_replication_redo != nullptr)
+      {
+	// this is the earliest it is ensured that no records are to be added anymore
+	m_parallel_replication_redo->wait_for_idle ();
+      }
+  }
+
+  void
+  replicator::redo_upto (cubthread::entry &thread_entry, const log_lsa &end_redo_lsa)
+  {
+    assert (m_redo_lsa < end_redo_lsa);
+
+    // redo all records from current position (m_redo_lsa) until end_redo_lsa
+
+    // make sure the log page is refreshed. otherwise it may be outdated and new records may be missed
+    m_reader.set_lsa_and_fetch_page (m_redo_lsa, log_reader::fetch_mode::FORCE);
+
+    while (m_redo_lsa < end_redo_lsa)
+      {
+	// read and redo a record
+	m_reader.set_lsa_and_fetch_page (m_redo_lsa);
+
+	const log_rec_header header = m_reader.reinterpret_copy_and_add_align<log_rec_header> ();
+
+	switch (header.type)
+	  {
+	  case LOG_REDO_DATA:
+	    read_and_redo_record<log_rec_redo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_MVCC_REDO_DATA:
+	    read_and_redo_record<log_rec_mvcc_redo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_UNDOREDO_DATA:
+	  case LOG_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record<log_rec_undoredo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_MVCC_UNDOREDO_DATA:
+	  case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record<log_rec_mvcc_undoredo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_RUN_POSTPONE:
+	    read_and_redo_record<log_rec_run_postpone> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_COMPENSATE:
+	    read_and_redo_record<log_rec_compensate> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_DBEXTERN_REDO_DATA:
+	  {
+	    log_rec_dbout_redo dbout_redo = m_reader.reinterpret_copy_and_add_align<log_rec_dbout_redo> ();
+	    log_rcv rcv;
+	    rcv.length = dbout_redo.length;
+	    log_rv_redo_record (&thread_entry, m_reader, RV_fun[dbout_redo.rcvindex].redofun, &rcv, &m_redo_lsa, 0,
+				nullptr, m_redo_unzip);
+	  }
+	  break;
+	  default:
+	    // do nothing
+	    break;
+	  }
+
+	{
+	  std::unique_lock<std::mutex> lock (m_redo_lsa_mutex);
+	  m_redo_lsa = header.forw_lsa;
+	}
+
+        // TODO: why is this notification inside this loop? where this function is
+        // called, can we be sure that 'm_redo_lsa == nxio_lsa' ?

```

Comment:
Either I did not understand or this notify is too early.

Reply:
Are you suggesting to move the notification to redo_upto_nxio_lsa? I think that would work too.

Reply:
Moved.